From 887f4d7232a7b80753e3d9e88c1ef86308e13247 Mon Sep 17 00:00:00 2001
From: Andrzej Surdej <Andrzej_Surdej@comcast.com>
Date: Tue, 13 Oct 2026 16:32:41 +0200
Subject: [PATCH] [MSE][GStreamer] Flush GStreamer pipeline when media element
 is detached from DOM

HTMLMediaElement::removedFromAncestor() now flushes the GStreamer
pipeline (and suspends the MSE source's data flow) instead of leaving
the MediaPlayer fully active while detached.
This releases media buffers and secure video memory for pages that detach
video elements without resetting or destroying them.
---
 Source/WebCore/html/HTMLMediaElement.cpp      | 25 +++++++++++-
 Source/WebCore/html/HTMLMediaElement.h        |  1 +
 .../WebCore/platform/graphics/MediaPlayer.cpp |  8 ++++
 .../WebCore/platform/graphics/MediaPlayer.h   |  3 ++
 .../platform/graphics/MediaPlayerPrivate.h    |  2 +
 .../mse/MediaPlayerPrivateGStreamerMSE.cpp    | 18 +++++++++
 .../mse/MediaPlayerPrivateGStreamerMSE.h      |  2 +
 .../mse/WebKitMediaSourceGStreamer.cpp        | 38 ++++++++++++++++---
 .../mse/WebKitMediaSourceGStreamer.h          |  3 ++
 9 files changed, 93 insertions(+), 7 deletions(-)

diff --git a/Source/WebCore/html/HTMLMediaElement.cpp b/Source/WebCore/html/HTMLMediaElement.cpp
index efaf5b8d3c52..ff91752befc5 100644
--- a/Source/WebCore/html/HTMLMediaElement.cpp
+++ b/Source/WebCore/html/HTMLMediaElement.cpp
@@ -1063,6 +1063,16 @@ void HTMLMediaElement::didFinishInsertingNode()
     if (m_inActiveDocument && m_networkState == NETWORK_EMPTY && !attributeWithoutSynchronization(srcAttr).isEmpty())
         prepareForLoad();
 
+#if PLATFORM(WPE)
+    if (RefPtr player = m_player) {
+        if (m_playerFlushedWhileDetached) {
+            m_playerFlushedWhileDetached = false;
+            seekInternal(currentMediaTime());
+            player->resumePlayback();
+        }
+    }
+#endif // PLATFORM(WPE)
+
     visibilityAdjustmentStateDidChange();
 
     if (!m_explicitlyMuted) {
@@ -1141,8 +1151,17 @@ void HTMLMediaElement::removedFromAncestor(RemovalType removalType, ContainerNod
     if (removalType.disconnectedFromDocument) {
         // Pause asynchronously to let the operation that removed us finish, in case we get inserted back into a document.
         queueTaskKeepingObjectAlive(*this, TaskSource::MediaElement, [this] {
-            if (!isContextStopped())
-                pauseAfterDetachedTask();
+            if (isContextStopped())
+                return;
+            pauseAfterDetachedTask();
+
+#if PLATFORM(WPE)
+            if (RefPtr player = m_player) {
+                ALWAYS_LOG(LOGIDENTIFIER, "Flushing media player pipeline of detached element");
+                player->flushPipelineAndSuspendPlayback();
+                m_playerFlushedWhileDetached = true;
+            }
+#endif // PLATFORM(WPE)
         });
     }
 
@@ -6634,6 +6653,7 @@ void HTMLMediaElement::clearMediaPlayer()
         player->invalidate();
         m_player = nullptr;
     }
+    m_playerFlushedWhileDetached = false;
     schedulePlaybackControlsManagerUpdate();
 
     stopPeriodicTimers();
@@ -7904,6 +7924,7 @@ void HTMLMediaElement::createMediaPlayer() WTF_IGNORES_THREAD_SAFETY_ANALYSIS
 #endif
 
     m_networkErrorOccured = false;
+    m_playerFlushedWhileDetached = false;
     m_lastContentTypeUsed.reset();
     if (RefPtr player = std::exchange(m_player, { })) {
         // The sniffer completionHandler would have taken a reference to the old MediaPlayer.
diff --git a/Source/WebCore/html/HTMLMediaElement.h b/Source/WebCore/html/HTMLMediaElement.h
index 417b3666ac87..0d5b2caa49d2 100644
--- a/Source/WebCore/html/HTMLMediaElement.h
+++ b/Source/WebCore/html/HTMLMediaElement.h
@@ -1417,6 +1417,7 @@ private:
 #endif
 
     bool m_isChangingReadyStateWhileSuspended { false };
+    bool m_playerFlushedWhileDetached { false };
     Atomic<unsigned> m_remainingReadyStateChangedAttempts;
 
     std::unique_ptr<PausableIntervalTimer> m_watchtimeTimer;
diff --git a/Source/WebCore/platform/graphics/MediaPlayer.cpp b/Source/WebCore/platform/graphics/MediaPlayer.cpp
index 9ef8f5ae9636..399b5099353a 100644
--- a/Source/WebCore/platform/graphics/MediaPlayer.cpp
+++ b/Source/WebCore/platform/graphics/MediaPlayer.cpp
@@ -1126,6 +1126,14 @@ void MediaPlayer::setPageIsSuspended(bool suspended)
     m_private->setPageIsSuspended(suspended);
 }
 
+void MediaPlayer::flushPipelineAndSuspendPlayback()
+{
+    m_private->flushPipelineAndSuspendPlayback();
+}
+void MediaPlayer::resumePlayback()
+{
+    m_private->resumePlayback();
+}
 void MediaPlayer::setVisibleForCanvas(bool visible)
 {
     if (visible == m_visibleForCanvas)
diff --git a/Source/WebCore/platform/graphics/MediaPlayer.h b/Source/WebCore/platform/graphics/MediaPlayer.h
index 5f21c445187d..7a12cfd6fb7c 100644
--- a/Source/WebCore/platform/graphics/MediaPlayer.h
+++ b/Source/WebCore/platform/graphics/MediaPlayer.h
@@ -425,6 +425,9 @@ public:
     using MediaPlayerEnums::BufferingPolicy;
     void setBufferingPolicy(BufferingPolicy);
 
+    void flushPipelineAndSuspendPlayback();
+    void resumePlayback();
+
 #if ENABLE(LEGACY_ENCRYPTED_MEDIA)
     // Represents synchronous exceptions that can be thrown from the Encrypted Media methods.
     // This is different from the asynchronous MediaKeyError.
diff --git a/Source/WebCore/platform/graphics/MediaPlayerPrivate.h b/Source/WebCore/platform/graphics/MediaPlayerPrivate.h
index d95cf96e33c9..c3a9be7af682 100644
--- a/Source/WebCore/platform/graphics/MediaPlayerPrivate.h
+++ b/Source/WebCore/platform/graphics/MediaPlayerPrivate.h
@@ -105,6 +105,8 @@ public:
     virtual void play() = 0;
     virtual void pause() = 0;
     virtual void setBufferingPolicy(MediaPlayer::BufferingPolicy) { }
+    virtual void flushPipelineAndSuspendPlayback() { }
+    virtual void resumePlayback() { }
 
     virtual bool supportsPictureInPicture() const { return false; }
     virtual bool supportsFullscreen() const { return false; }
diff --git a/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp b/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp
index 958059298618..b54193122236 100644
--- a/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp
@@ -536,6 +536,24 @@ void MediaPlayerPrivateGStreamerMSE::setInitialVideoSize(const FloatSize& videoS
     m_videoSize = videoSize;
 }
 
+void MediaPlayerPrivateGStreamerMSE::flushPipelineAndSuspendPlayback()
+{
+    if (!m_source)
+        return;
+
+    GST_INFO_OBJECT(pipeline(), "Suspending and flushing all WebKitMediaSrc streams");
+    webKitMediaSrcFlushAndSuspend(webKitMediaSrc());
+}
+
+void MediaPlayerPrivateGStreamerMSE::resumePlayback()
+{
+    if (!m_source)
+        return;
+
+    GST_INFO_OBJECT(pipeline(), "Resuming all WebKitMediaSrc streams");
+    webKitMediaSrcResume(webKitMediaSrc());
+}
+
 void MediaPlayerPrivateGStreamerMSE::startSource(const Vector<RefPtr<MediaSourceTrackGStreamer>>& tracks)
 {
     m_tracks = filterOutRepeatingTracks(tracks);
diff --git a/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h b/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h
index 45b12dd68aca..4d945614a3e4 100644
--- a/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h
+++ b/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h
@@ -67,6 +67,8 @@ public:
     MediaTime maxTimeSeekable() const override;
     bool timeIsProgressing() const override;
     void notifyActiveSourceBuffersChanged() final;
+    void flushPipelineAndSuspendPlayback() final;
+    void resumePlayback() final;
 
     void sourceSetup(GstElement*) override;
 
diff --git a/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp b/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp
index 3ea25ff8e38d..9fd53600a9fb 100644
--- a/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp
@@ -103,7 +103,7 @@ static gboolean webKitMediaSrcActivateMode(GstPad*, GstObject*, GstPadMode, gboo
 static void webKitMediaSrcLoop(void*);
 static void webKitMediaSrcTearDownStream(WebKitMediaSrc* source, TrackID);
 static void webKitMediaSrcGetProperty(GObject*, unsigned propId, GValue*, GParamSpec*);
-static void webKitMediaSrcStreamFlush(Stream*, bool isSeekingFlush);
+static void webKitMediaSrcStreamFlush(Stream*, bool isSeekingFlush, bool suspend = false);
 static gboolean webKitMediaSrcSendEvent(GstElement*, GstEvent*);
 static RefPtr<MediaPlayerPrivateGStreamerMSE> webKitMediaSrcPlayer(WebKitMediaSrc*);
 
@@ -651,7 +651,7 @@ static void webKitMediaSrcLoop(void* userData)
         ASSERT_NOT_REACHED();
 }
 
-static void webKitMediaSrcStreamFlush(Stream* stream, bool isSeekingFlush)
+static void webKitMediaSrcStreamFlush(Stream* stream, bool isSeekingFlush, bool suspend)
 {
     ASSERT(isMainThread());
     bool skipFlush = false;
@@ -661,7 +661,7 @@ static void webKitMediaSrcStreamFlush(Stream* stream, bool isSeekingFlush)
     {
         DataMutexLocker streamingMembers { stream->streamingMembersDataMutex };
 
-        if (!streamingMembers->hasPoppedFirstObject) {
+        if (!streamingMembers->hasPoppedFirstObject && !suspend) {
             GST_DEBUG_OBJECT(stream->source, "Flush request for stream '%" PRIu64 "' occurred before hasPoppedFirstObject, just clearing the queue and readjusting the segment.", stream->track->id());
             DataMutexLocker queue { stream->track->queueDataMutex() };
             // We use clear() instead of flush() because the WebKitMediaSrc streaming thread could be waiting
@@ -763,8 +763,12 @@ static void webKitMediaSrcStreamFlush(Stream* stream, bool isSeekingFlush)
             streamingMembers->hasPoppedFirstObject = false;
         }
 
-        GST_DEBUG_OBJECT(stream->pad.get(), "Starting webKitMediaSrcLoop task and releasing the STREAM_LOCK.");
-        gst_pad_start_task(stream->pad.get(), webKitMediaSrcLoop, stream->pad.get(), nullptr);
+        if (!suspend) {
+            GST_DEBUG_OBJECT(stream->pad.get(), "Starting webKitMediaSrcLoop task and releasing the STREAM_LOCK.");
+            gst_pad_start_task(stream->pad.get(), webKitMediaSrcLoop, stream->pad.get(), nullptr);
+        } else {
+            GST_INFO_OBJECT(stream->pad.get(), "Data flow is suspended, not starting webKitMediaSrcLoop task.");
+        }
     }
 
     GST_DEBUG_OBJECT(stream->source, "Flush request for stream '%" PRIu64 "' (isSeekingFlush = %s) satisfied.",
@@ -785,6 +789,30 @@ void webKitMediaSrcFlush(WebKitMediaSrc* source, TrackID streamId)
     webKitMediaSrcStreamFlush(stream, false);
 }
 
+void webKitMediaSrcFlushAndSuspend(WebKitMediaSrc* source)
+{
+    ASSERT(isMainThread());
+    GST_INFO_OBJECT(source, "Flushing all streams and suspending data flow.");
+
+    // Flush all streams and pause streaming threads.
+    // The streaming threads will remain paused until webKitMediaSrcResume is called,
+    // hasPoppedFirstObject will prevent the streaming threads from resuming until explicitly requested.
+    for (const RefPtr<Stream>& stream : source->priv->streams.values()) {
+        webKitMediaSrcStreamFlush(stream.get(), false, true);
+    }
+}
+
+void webKitMediaSrcResume(WebKitMediaSrc* source)
+{
+    ASSERT(isMainThread());
+    GST_INFO_OBJECT(source, "Resuming data flow for all streams.");
+
+    for (const RefPtr<Stream>& stream : source->priv->streams.values()) {
+        ASSERT(stream->pad);
+        gst_pad_start_task(stream->pad.get(), webKitMediaSrcLoop, stream->pad.get(), nullptr);
+    }
+}
+
 static void webKitMediaSrcSeek(WebKitMediaSrc* source, uint64_t startTime, double rate)
 {
     ASSERT(isMainThread());
diff --git a/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.h b/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.h
index a24412b2c5cf..271f453245e1 100644
--- a/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.h
+++ b/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.h
@@ -65,6 +65,9 @@ void webKitMediaSrcEmitStreams(WebKitMediaSrc* source, const Vector<RefPtr<WebCo
 
 void webKitMediaSrcFlush(WebKitMediaSrc*, TrackID streamId);
 
+void webKitMediaSrcFlushAndSuspend(WebKitMediaSrc*);
+void webKitMediaSrcResume(WebKitMediaSrc*);
+
 void webKitMediaSrcSetPlayer(WebKitMediaSrc*, ThreadSafeWeakPtr<WebCore::MediaPlayerPrivateGStreamerMSE>&&);
 
 G_END_DECLS
-- 
2.53.0

